test/test_ops.py

```diff
@@ -598,8 +600,160 @@ def test_vjpvmap(self, device, dtype, op):
+        if is_inplace(op, op.get_op()):
+            self.skipTest("op is inplace")
+            return
+        print(op.op)
+        _requires_grad = op.supports_autograd and dtype.is_floating_point
+        # print(dtype)
+
+        # print(_requires_grad)
+        samples = op.sample_inputs(device, dtype, requires_grad=_requires_grad)
+        # Acquires variants to test
+        def wrap_tensor(x):
+            if type(x) == torch.Tensor:
+                return DecompositionTensor(x)
+            return x
+
+        def unwrap_tensor(x):
+            if type(x) == DecompositionTensor:
+                if not hasattr(x, 'elem'):
+                    raise InplaceError()
+                return x.elem
+            return x
+
+        def _assertEqual(a, b):
+            if dtype == torch.half:
+                self.assertEqual(a, b, rtol=0.001, atol=1e-04)
+            else:
+                self.assertEqual(a, b)
+
+        try:
+            func = op.get_op()
+            for sample_input in samples:
+                if _requires_grad:
+                    fn, primals = normalize_op_for_vjp(func, sample_input)
+                    result = fn(*primals)
+                    out, vjp_fn = ref_vjp(fn, *primals)
+                    cotangents = tree_map(lambda x: torch.randn_like(x), out)
+                    expected_grads = vjp_fn(cotangents)
+
+                    decomp_out, decomp_vjp_fn = ref_vjp(fn, *tree_map(wrap_tensor, primals))
+                    _assertEqual(out, tree_map(unwrap_tensor, decomp_out))
+
+                    decomp_grads = decomp_vjp_fn(cotangents)
+                    _assertEqual(expected_grads, tree_map(unwrap_tensor, decomp_grads))
+
+                else:
+                    args = [sample_input.input] + list(sample_input.args)
+                    kwargs = sample_input.kwargs
+                    orig_out = func(*args, **kwargs)
+
+                    args = tree_map(wrap_tensor, args)
+                    kwargs = tree_map(wrap_tensor, kwargs)
+                    decomp_out = func(*args, **kwargs)
+                    _assertEqual(orig_out, tree_map(unwrap_tensor, decomp_out))
+
+
+        except InplaceError:
+            self.skipTest("op is inplace")
+            return
+        except RuntimeError as e:
+            if "not implemented for 'Half'" in str(e):
+                self.skipTest("(backwards) op not supported for Half")
+                return
+            if "Mismatch in shape: grad_output" in str(e):
+                self.skipTest("Some weird issue with autograd engine and tensor subclasses")
+            raise e
+
+    def test_placeholder(self):
+        with open('op_analysis/run_ops.txt', 'w') as f:
+            for op in DecompositionTensor.run_ops:
+                f.write(f'{op.__name__}\n')
+        with open('op_analysis/run_decompositions.txt', 'w') as f:
+            for op in DecompositionTensor.run_decompositions:
+                f.write(f'{op.__name__}\n')
+
```

Comment:
What's this for?

Reply:
I keep track of which ops are run in the OpInfo tests and which decompositions are run so that I can dump them out to a file, which gets loaded here: https://docs.google.com/spreadsheets/d/1Sp4HUjxwMifS5oDQg0yvjqk7hKOpCfKO4jWH4MTGP-k/edit#gid=1815506345